okhttp-clients leaks response bodies when calls are interrupted

The `PaxosQuorumChecker` in atlasdb makes remote calls to multiple remote `PaxosAcceptor.getLatestSequencePreparedOrAccepted` endpoints, waits for a quorum of them to respond, then attempts to cancel the remainder. This appears to result in a leak of an okhttp3 `ResponseBody`
```
A connection to <REDACTED> was leaked. Did you forget to close a response body?
	at okhttp3.internal.platform.Platform.getStackTraceForCloseable(Platform.java:149)
	at okhttp3.RealCall.captureCallStackTrace(RealCall.java:116)
	at okhttp3.RealCall.enqueue(RealCall.java:125)
	at com.palantir.conjure.java.okhttp.ForwardingCall.enqueue(ForwardingCall.java:48)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall.enqueueInternal(RemotingOkHttpCall.java:217)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall.enqueueClosingEntireSpan(RemotingOkHttpCall.java:201)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall.access$100(RemotingOkHttpCall.java:59)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall$2.onSuccess(RemotingOkHttpCall.java:187)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall$2.onSuccess(RemotingOkHttpCall.java:181)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1021)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.ImmediateFuture.addListener(ImmediateFuture.java:38)
	at com.google.common.util.concurrent.Futures.addCallback(Futures.java:996)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall.enqueue(RemotingOkHttpCall.java:181)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall.execute(RemotingOkHttpCall.java:114)
	at feign.okhttp.OkHttpClient.execute(OkHttpClient.java:160)
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:98)
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:77)
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:102)
	at com.sun.proxy.$Proxy83.getLatestSequencePreparedOrAccepted(Unknown Source)
	at com.palantir.tritium.proxy.InstrumentedPaxosAcceptor$4.getLatestSequencePreparedOrAccepted(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor117.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.palantir.common.proxy.ExperimentRunningProxy.handleInvocation(ExperimentRunningProxy.java:91)
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:86)
	at com.sun.proxy.$Proxy83.getLatestSequencePreparedOrAccepted(Unknown Source)
	at com.palantir.paxos.SingleLeaderAcceptorNetworkClient.lambda$getLatestSequencePreparedOrAccepted$2(SingleLeaderAcceptorNetworkClient.java:64)
	at com.palantir.paxos.PaxosQuorumChecker.lambda$collectResponses$4(PaxosQuorumChecker.java:196)
	at com.palantir.common.concurrent.MultiplexingCompletionService.lambda$submitAndPrepareForQueueing$0(MultiplexingCompletionService.java:88)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:176)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.palantir.common.concurrent.PTExecutors.lambda$wrapRunnable$1(PTExecutors.java:602)
	at com.palantir.tracing.Tracers$TracingAwareRunnable.run(Tracers.java:492)
	at com.palantir.common.concurrent.ForwardingRunnableFuture.run(ForwardingRunnableFuture.java:60)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
The call in atlasdb appears to swallow any actual error produced by the remote call itself, so I'm not sure how the response body is actually being leaked.

I made an attempt to reproduce this in a simple test that simply made remote requests in a loop and cancelled them, but was unable to reproduce this error. Instead, I got a different error logged
```
com.palantir.conjure.java.okhttp.LeakDetector: Resource leak detected - did you forget to close a resource properly? This will likely hurt performance. Stack trace is of the call where the acquire happened. (resourceType: com.netflix.concurrency.limits.Limiter$Listener)
com.palantir.logsafe.exceptions.SafeRuntimeException: Runtime exception for stack trace
	at com.palantir.conjure.java.okhttp.LeakDetector.maybeCreateStackTrace(LeakDetector.java:38)
	at com.palantir.conjure.java.okhttp.ConcurrencyLimiters$DefaultConcurrencyLimiter.acquire(ConcurrencyLimiters.java:231)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall.enqueue(RemotingOkHttpCall.java:179)
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall.execute(RemotingOkHttpCall.java:114)
	at feign.okhttp.OkHttpClient.execute(OkHttpClient.java:160)
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:98)
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:77)
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:102)
```
along with this error from the remote call
```
feign.RetryableException: Call cancelled via interruption executing GET <REDACTED>
	at feign.FeignException.errorExecuting(FeignException.java:65)
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:105)
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:77)
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:102)
	at com.sun.proxy.$Proxy25.getVersion(Unknown Source)
Caused by: java.io.InterruptedIOException: Call cancelled via interruption
	at com.palantir.conjure.java.okhttp.RemotingOkHttpCall.execute(RemotingOkHttpCall.java:138)
	at feign.okhttp.OkHttpClient.execute(OkHttpClient.java:160)
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:98)
	... 8 more
```

--- a/changelog/@unreleased/pr-1415.v2.yml
+++ b/changelog/@unreleased/pr-1415.v2.yml
@@ -1,0 +1,6 @@
+type: fix
+fix:
+  description: Fix a response leak when requests are canceled. This is a workaround
+    for an upstream bug in okhttp.
+  links:
+  - https://github.com/palantir/conjure-java-runtime/pull/1415

--- a/okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/OkHttpClients.java
+++ b/okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/OkHttpClients.java
@@ -212,6 +212,9 @@ public final class OkHttpClients {
 
         client.addInterceptor(CatchThrowableInterceptor.INSTANCE);
         client.addInterceptor(SpanTerminatingInterceptor.INSTANCE);
+        // Order is important, this interceptor must be applied prior to ConcurrencyLimitingInterceptor
+        // in order to prevent concurrency limiters from leaking.
+        client.addInterceptor(ResponseCapturingInterceptor.INSTANCE);
 
         // Routing
         UrlSelectorImpl urlSelector = UrlSelectorImpl.createWithFailedUrlCooldown(

--- a/okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java
+++ b/okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java
@@ -242,7 +242,7 @@ final class RemotingOkHttpCall extends ForwardingCall {
     }
 
     private void enqueueInternal(Callback callback) {
-        super.enqueue(new Callback() {
+        super.enqueue(new LeakedResponseClosingCallback(new Callback() {
             @Override
             public void onFailure(Call call, IOException exception) {
                 if (isCanceled()) {
@@ -352,7 +352,29 @@ final class RemotingOkHttpCall extends ForwardingCall {
                         new SafeIoException(
                                 "Failed to handle request, this is an conjure-java-runtime bug."));
             }
-        });
+        }));
+    }
+
+    private static final class LeakedResponseClosingCallback implements Callback {
+
+        private final Callback delegate;
+
+        LeakedResponseClosingCallback(Callback delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onFailure(Call call, IOException exception) {
+            ResponseCapturingInterceptor.getResponse().ifPresent(RemotingOkHttpCall::close);
+            ResponseCapturingInterceptor.clearThreadState();
+            delegate.onFailure(call, exception);
+        }
+
+        @Override
+        public void onResponse(Call call, Response response) throws IOException {
+            ResponseCapturingInterceptor.clearThreadState();
+            delegate.onResponse(call, response);
+        }
     }
 
     private boolean shouldRetry(IOException exception, Optional<Duration> backoff) {

--- a/okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ResponseCapturingInterceptor.java
+++ b/okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ResponseCapturingInterceptor.java
@@ -1,0 +1,47 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.okhttp;
+
+import java.io.IOException;
+import java.util.Optional;
+import okhttp3.Interceptor;
+import okhttp3.Response;
+
+/**
+ * Workaround a bug in okhttp where cancellation in flight results in leaked responses.
+ * https://github.com/square/okhttp/blob/d28d2cec21641b61f3d34e05dd52f43a717c2d32/okhttp/src/main/java/okhttp3/RealCall.java#L210-L213
+ */
+enum ResponseCapturingInterceptor implements Interceptor {
+    INSTANCE;
+
+    private static final ThreadLocal<Response> currentResponse = new ThreadLocal<>();
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Response response = chain.proceed(chain.request());
+        currentResponse.set(response);
+        return response;
+    }
+
+    static Optional<Response> getResponse() {
+        return Optional.ofNullable(currentResponse.get());
+    }
+
+    static void clearThreadState() {
+        currentResponse.remove();
+    }
+}

